sys/include/net/nanocoap.h

```diff
@@ -347,6 +347,20 @@ void *coap_request_ctx_get_context(const coap_request_ctx_t *ctx);
  */
 uint32_t coap_request_ctx_get_tl_type(const coap_request_ctx_t *ctx);
 
+#if defined(MODULE_SOCK_UDP) || DOXYGEN
+#include "net/sock/udp.h"
+/**
+ * @brief   Get the remote endpoint from which the request was received
+ *
+ * @note    This is currently only implemented for GCoAP
+ *
+ * @param[in]   ctx The request context
+ *
+ * @return  Remote endpoint from which the request originated
+ */
+const sock_udp_ep_t *coap_request_ctx_get_remote(const coap_request_ctx_t *ctx);
```

Comment:
```suggestion
/**
 * @brief   Get the remote endpoint from which the request was received
 *
 * @note    This is currently only implemented for GCoAP
 *
 * @param[in]   ctx The request context
 *
 * @return  Remote endpoint from which the request originated
 * @retval  NULL    The request was not received via UDP
 */
const sock_udp_ep_t *coap_request_ctx_get_remote_udp(const coap_request_ctx_t *ctx);
```

How about this? Theoretically, CoAP could be transported over UDP, DTLS, OSCORE, TCP, WebSocket, and (not yet standardized) QUIC. Not all are compatible with `sock_udp_ep_t`, so maybe adding a postfix would be a simple way to eventually support more transports?

Otherwise, this looks good to me :)

@fabian18 has a related PR open that allows accessing the local address. If I am not mistaken, that is still not yet in, right?

@fabian18: Would this API match the one for accessing the local endpoint? It would IMO be a pity, if those two APIs would be inconsistent.

Reply:
I see #16827 has been closed by stalebot. This PR is the better way to go, since we have `coap_request_ctx_t` now.

> Not all are compatible with sock_udp_ep_t

I would have expected something that leaves the possibility to encode more transport layer types.
For example:
```C
typedef union {
#if IS_USED(MODULE_SOCK_UDP)
    sock_udp_ep_t *udp;
#endif
#if IS_USED(MODULE_SOCK_TCP)
    sock_tcp_ep_t *tcp;
#endif
} coap_ep_t;
```
I would have put two `coap_ep_t` into `coap_request_ctx_t`, for local and remote endpoint information, and encode the type in `tl_type`. And then the internal API needs to pass on the information to the point where a  `coap_request_ctx_t` is built.

Reply:
Having both local and remote is valuable (eg. when handling URIs and when handling multicast). Whether or not these always take pair shape is up to the transport. (For example, in non-group OSCORE, there wouldn't be a pointer per end, but merely a single pointer that describes the OSCORE context that was used, without any two distinct things to point at).

Thus, maybe the return value could be something like

```c
typedef struct {
    enum coap_request_ctx_remote_type;
    union {
#ifdef ...
        struct {
            sock_udp_ep_t *remote;
            sock_udp_ep_t *local;
        } udp;
#endif
#ifdef ...
        struct {
            sock_tcp_ep_t *remote;
            sock_tcp_ep_t *local;
        } tcp;
#endif
    };
} coap_request_ctx_remote_t;
```

I think that that's still compact enough to just return as a struct; it shouldn't be a pointer into the CoAP stack (eg. the memo) because that'd be limiting through API stability.

---

Alternatively, if we don't want to go the route of complex structs, the function could be declared UDP-only (returning NULL in non-UDP cases), and there'd be an accessor for each transport individually. A handler would try getting any address it understands, and immediately use results when they are available.

I'm good with either.

Reply:
I think a transport specific getter function is easier to handle.
Returning a complex struct would just blow up the API, returning a simple type is more ergonomic and easier to keep compatible in the future.

Reply:
This also neatly sidesteps the topic of a local address. A `coap_request_ctx_get_local_udp` function could still be introduced compatibly and in style, when the need for them comes up.

If non-UDP protocols come up later, they may also (instead of getting the strictly typed) have the URI (or CRI, #13827) of the local or remote side built, and *that* would then be transport independent.